### PR TITLE
bpf: BPF Host Routing supports lookup linux custom FIB rules in multi-device environment

### PR DESCRIPTION
**Pull request by weiwei02**
bpf: BPF Host Routing supports lookup linux custom FIB rules in multi-device environment

the FIB rules of linux will be ignored during fast redirection in the
multi-device scenario, and only use the main route table to lookup
route entries. This routing query method seems to improve the
performance of routing query, but it also occurs some problems:

1. different data paths cannot pass the reverse path filter. For
example, in the BaiDu Cloud, there are multiple ENIs on a machine,
and each ENI is associated with multiple Pods., Pod IP is the secondary
 IP of the ENI. Packets originating from Pod should be transmitted
from the ENI associated with Pod, not the default netlink of the host.

The previous pr has lift the restriction of fast redirect on multiple
devices, but it has not lift the restriction of linux custom route
rules. see

Fixes: 92edccfad0 (bpf: lift multi dev restriction on fast redirect)

### CODE REVIEW

File: daemon/cmd/daemon_main.go
Comment (pippolo84) [inline daemon/cmd/daemon_main.go]: Nit: following [Go naming conventions](https://github.com/golang/go/wiki/CodeReviewComments#initialisms) for acronyms it should be

```suggestion
	flags.Bool(option.EnableBPFFIBCustomRules, defaults.EnableBPFFIBCustomRules, "Use custom fib rule lookup for BPF Host Routing multi-device deployments")
	option.BindEnv(Vp, option.EnableBPFFIBCustomRules)

```

**Review by pippolo84 — CHANGES_REQUESTED** [review]
Thanks for the PR!

Overall looks good, apart a nit regarding naming conventions.
Also, I suggest to improve the commit message, wrapping lines at 72 characters and using:

`Fixes: 92edccfad0 (bpf: lift multi dev restriction on fast redirect)`

to reference the previous related commit.

File: daemon/cmd/daemon_main.go
Comment (weiwei02) [reply daemon/cmd/daemon_main.go]: Thank you. I will modify the code according to good coding suggestions

**Review by pippolo84 — APPROVED** [review]
LGTM, thanks! :100:

**Comment by borkmann** [comment]
Hi @weiwei02, is it possible to rebase this work on top of https://github.com/cilium/cilium/pull/23884 ? I've been reworking all fib lookup locations to reuse common fib code there.

**Comment by weiwei02** [comment]
> Hi @weiwei02, is it possible to rebase this work on top of #23884 ? I've been reworking all fib lookup locations to reuse common fib code there.

@borkmann I saw your refactoring, which makes the code more concise. I agree with your idea of rebase， but I just don't know how to rabase for 'open' pr .Or wait until the pull request is merged, and then rebase it to the top of #23884

**Comment by borkmann** [comment]
@weiwei02 done now.

**Comment by weiwei02** [comment]
> Hi @weiwei02, is it possible to rebase this work on top of #23884 ? I've been reworking all fib lookup locations to reuse common fib code there.

@pippolo84 @borkmann I have rebase the code that supports custom FIB rules based on #23884. Please help me review again, thanks.

**Review by pippolo84 — APPROVED** [review]
LGTM, thanks! :rocket:

File: pkg/datapath/linux/config/config.go
Comment (borkmann) [inline pkg/datapath/linux/config/config.go]: I would just define a `FIB_LOOKUP_FLAGS` right here, and if option.Config.EnableBPFFIBCustomRules is enabled set this to 0, otherwise "BPF_FIB_LOOKUP_DIRECT" string. Then you don't need the whole ENABLE_BPF_FIB_CUSTOM_RULES detour, and fib_lookup() will use FIB_LOOKUP_FLAGS directly.

**Review by borkmann — CHANGES_REQUESTED** [review]
(rest lgtm)

File: pkg/datapath/linux/config/config.go
Comment (weiwei02) [reply pkg/datapath/linux/config/config.go]: good idea!

**Comment by weiwei02** [comment]
> (rest lgtm)

@borkmann Can you help me review the code again, the  change requested has been completed.

**Review by ldelossa — APPROVED** [review]
changes look OK to me.

**Comment by borkmann** [comment]
Hi @weiwei02, sry for late reply, the issue was fixed some time ago in master by unconditionally doing full fib lookup as part of a bug fix. This actually also addresses your case. Now, opting-in for a direct fib lookup could be something to consider, but at the same time, I'm not sure if it's worth it, certainly needs some concrete benchmark numbers as a base for discussion first. I would suggest we could mark this one here as resolved/closed, unless you would like to pursue the opt-in approach from your side then we can always reopen. Thanks again!

**Pull request closed without merging**
